<a href="https://colab.research.google.com/github/holegar/datasharing/blob/master/Numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Week 8 Exercises
Let's practice what we have learnt on numpy library

import NumPy as np

In [ ]:
import numpy as np

In [ ]:
b = np.arange(1,28)

In [ ]:
b = b.reshape(3,3,3)

In [ ]:
b

array([[[ 1,  2,  3],
        [ 4,  5,  6],
        [ 7,  8,  9]],

       [[10, 11, 12],
        [13, 14, 15],
        [16, 17, 18]],

       [[19, 20, 21],
        [22, 23, 24],
        [25, 26, 27]]])

In [ ]:
b[0,1,1]

5

In [ ]:
Create an array of 20 zeros

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.zeros(20)

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0.])

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0.])

In [ ]:
Create an array of 20 ones

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.ones(20)

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1.])

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1.])

Create an array of 20 threes

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
[3]*20

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

array([3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3.])

In [ ]:
Create an array of integers from 5 to 100

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.arange(5,100)

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55,
       56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72,
       73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,
       90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

array([  5,   6,   7,   8,   9,  10,  11,  12,  13,  14,  15,  16,  17,
        18,  19,  20,  21,  22,  23,  24,  25,  26,  27,  28,  29,  30,
        31,  32,  33,  34,  35,  36,  37,  38,  39,  40,  41,  42,  43,
        44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  55,  56,
        57,  58,  59,  60,  61,  62,  63,  64,  65,  66,  67,  68,  69,
        70,  71,  72,  73,  74,  75,  76,  77,  78,  79,  80,  81,  82,
        83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,  94,  95,
        96,  97,  98,  99, 100])

create an array of even numbers from 10 to 1000

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.arange(10,1000,2)

array([ 10,  12,  14,  16,  18,  20,  22,  24,  26,  28,  30,  32,  34,
        36,  38,  40,  42,  44,  46,  48,  50,  52,  54,  56,  58,  60,
        62,  64,  66,  68,  70,  72,  74,  76,  78,  80,  82,  84,  86,
        88,  90,  92,  94,  96,  98, 100, 102, 104, 106, 108, 110, 112,
       114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138,
       140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164,
       166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190,
       192, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216,
       218, 220, 222, 224, 226, 228, 230, 232, 234, 236, 238, 240, 242,
       244, 246, 248, 250, 252, 254, 256, 258, 260, 262, 264, 266, 268,
       270, 272, 274, 276, 278, 280, 282, 284, 286, 288, 290, 292, 294,
       296, 298, 300, 302, 304, 306, 308, 310, 312, 314, 316, 318, 320,
       322, 324, 326, 328, 330, 332, 334, 336, 338, 340, 342, 344, 346,
       348, 350, 352, 354, 356, 358, 360, 362, 364, 366, 368, 37

array([  10,   12,   14,   16,   18,   20,   22,   24,   26,   28,   30,
         32,   34,   36,   38,   40,   42,   44,   46,   48,   50,   52,
         54,   56,   58,   60,   62,   64,   66,   68,   70,   72,   74,
         76,   78,   80,   82,   84,   86,   88,   90,   92,   94,   96,
         98,  100,  102,  104,  106,  108,  110,  112,  114,  116,  118,
        120,  122,  124,  126,  128,  130,  132,  134,  136,  138,  140,
        142,  144,  146,  148,  150,  152,  154,  156,  158,  160,  162,
        164,  166,  168,  170,  172,  174,  176,  178,  180,  182,  184,
        186,  188,  190,  192,  194,  196,  198,  200,  202,  204,  206,
        208,  210,  212,  214,  216,  218,  220,  222,  224,  226,  228,
        230,  232,  234,  236,  238,  240,  242,  244,  246,  248,  250,
        252,  254,  256,  258,  260,  262,  264,  266,  268,  270,  272,
        274,  276,  278,  280,  282,  284,  286,  288,  290,  292,  294,
        296,  298,  300,  302,  304,  306,  308,  3

In [ ]:
Create a 5 by 5 matrix with values ranging between 0 and 10

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.random.randint(0,10,25).reshape(5,5)

array([[3, 3, 8, 3, 9],
       [6, 0, 5, 6, 6],
       [2, 9, 1, 2, 6],
       [0, 4, 0, 4, 0],
       [2, 2, 3, 3, 6]])

array([[ 0,  7,  6,  4,  0],
       [ 7, 10,  2,  7,  0],
       [ 1,  0, 10,  7,  2],
       [ 3,  9,  3, 10,  9],
       [ 9,  8,  2, 10,  8]])

Create a 5 by 5 identity matrix

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.eye(5)

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

array([[1., 0., 0., 0., 0.],
       [0., 1., 0., 0., 0.],
       [0., 0., 1., 0., 0.],
       [0., 0., 0., 1., 0.],
       [0., 0., 0., 0., 1.]])

Generate an array of 250 random numbers sampled from standard normal distribution 

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.random.randint()

array([[  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
         27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
         40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50],
       [ 51,  52,  53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,
         64,  65,  66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,
         77,  78,  79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,
         90,  91,  92,  93,  94,  95,  96,  97,  98,  99, 100],
       [101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113,
        114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126,
        127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139,
        140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150],
       [151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163,
        164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176,
     

array([-1.26172415e+00, -1.88004913e+00, -1.78221646e-03, -9.49368373e-01,
       -1.28635291e+00, -1.16684873e-01,  1.09314519e-01, -1.44235016e+00,
       -2.50381281e+00, -1.20626710e-01, -8.23380413e-01,  1.58411253e+00,
        6.73867082e-01,  3.05541045e-01,  6.41824429e-01, -4.97812470e-01,
        1.03882111e+00,  8.35774515e-01, -4.42634475e-01,  3.21961582e-01,
       -1.04939805e+00, -1.35675983e+00, -9.78392088e-02,  6.94794785e-01,
       -1.72403541e+00,  6.76724239e-01, -1.12049796e+00, -1.42832141e+00,
        3.45615137e-01,  2.31570753e+00, -3.22010051e-01, -9.36094029e-01,
       -1.12869733e+00, -1.51811112e+00,  5.31068142e-01, -5.49224672e-01,
        9.88034848e-01,  1.57788504e+00, -2.35876171e-01, -1.02231714e+00,
        1.28647825e+00, -6.30679426e-01, -1.03682363e-02, -6.23531394e-01,
        1.25841584e-01,  9.67428245e-01,  1.65107505e-03,  7.75764219e-01,
       -1.12573737e-01,  1.49043785e+00, -3.21087399e-01,  2.80119340e-01,
       -8.25276617e-01, -

Create the following matrix

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
np.linspace(0,2,50)

array([0.        , 0.04081633, 0.08163265, 0.12244898, 0.16326531,
       0.20408163, 0.24489796, 0.28571429, 0.32653061, 0.36734694,
       0.40816327, 0.44897959, 0.48979592, 0.53061224, 0.57142857,
       0.6122449 , 0.65306122, 0.69387755, 0.73469388, 0.7755102 ,
       0.81632653, 0.85714286, 0.89795918, 0.93877551, 0.97959184,
       1.02040816, 1.06122449, 1.10204082, 1.14285714, 1.18367347,
       1.2244898 , 1.26530612, 1.30612245, 1.34693878, 1.3877551 ,
       1.42857143, 1.46938776, 1.51020408, 1.55102041, 1.59183673,
       1.63265306, 1.67346939, 1.71428571, 1.75510204, 1.79591837,
       1.83673469, 1.87755102, 1.91836735, 1.95918367, 2.        ])

array([[ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ],
       [ 5.1,  5.2,  5.3,  5.4,  5.5,  5.6,  5.7,  5.8,  5.9,  6. ,  6.1,
         6.2,  6.3,  6.4,  6.5,  6.6,  6.7,  6.8,  6.9,  7. ,  7.1,  7.2,
         7.3,  7.4,  7.5,  7.6,  7.7,  7.8,  7.9,  8. ,  8.1,  8.2,  8.3,
         8.4,  8.5,  8.6,  8.7,  8.8,  8.9,  9. ,  9.1,  9.2,  9.3,  9.4,
         9.5,  9.6,  9.7,  9.8,  9.9, 10. ],
       [10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. , 11.1,
        11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1, 12.2,
        12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2, 13.3,
        13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3, 14.4,
        14.5, 14.6, 14

In [ ]:
Let's generate an array of 50 linearly spaced elements between 0 and 1

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below

array([0.        , 0.02040816, 0.04081633, 0.06122449, 0.08163265,
       0.10204082, 0.12244898, 0.14285714, 0.16326531, 0.18367347,
       0.20408163, 0.2244898 , 0.24489796, 0.26530612, 0.28571429,
       0.30612245, 0.32653061, 0.34693878, 0.36734694, 0.3877551 ,
       0.40816327, 0.42857143, 0.44897959, 0.46938776, 0.48979592,
       0.51020408, 0.53061224, 0.55102041, 0.57142857, 0.59183673,
       0.6122449 , 0.63265306, 0.65306122, 0.67346939, 0.69387755,
       0.71428571, 0.73469388, 0.75510204, 0.7755102 , 0.79591837,
       0.81632653, 0.83673469, 0.85714286, 0.87755102, 0.89795918,
       0.91836735, 0.93877551, 0.95918367, 0.97959184, 1.        ])

In [ ]:
#Numpy Indexing and Slicing using the provided matrices. You need to replicate the matrix's outputs

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below

In [ ]:
d_mat = np.arange(1,51).reshape(5,10)
d_mat

array([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25, 26, 27, 28, 29, 30],
       [31, 32, 33, 34, 35, 36, 37, 38, 39, 40],
       [41, 42, 43, 44, 45, 46, 47, 48, 49, 50]])

In [ ]:
#Slice the array to produce the array below  #Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat[2:,1:]

array([[22, 23, 24, 25, 26, 27, 28, 29, 30],
       [32, 33, 34, 35, 36, 37, 38, 39, 40],
       [42, 43, 44, 45, 46, 47, 48, 49, 50]])

array([[22, 23, 24, 25, 26, 27, 28, 29, 30],
       [32, 33, 34, 35, 36, 37, 38, 39, 40],
       [42, 43, 44, 45, 46, 47, 48, 49, 50]])

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat[2,7]

28

28

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat[:4,2:3]

array([[ 3],
       [13],
       [23],
       [33]])

array([[ 3],
       [13],
       [23],
       [33]])

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat[3,:]

array([31, 32, 33, 34, 35, 36, 37, 38, 39, 40])

array([31, 32, 33, 34, 35, 36, 37, 38, 39, 40])

In [ ]:
#compute the sum of all elements of d_mat

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.sum()

1275

1275

In [ ]:
#Compute the standard deviation 

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.std()

14.430869689661812

14.430869689661812

In [ ]:
#compute the mean

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.mean()

25.5

25.5

In [ ]:
#compute the variance

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.var()

208.25

208.25

In [ ]:
#Compute the sum of all columns

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.sum(axis=0)

array([105, 110, 115, 120, 125, 130, 135, 140, 145, 150])

array([105, 110, 115, 120, 125, 130, 135, 140, 145, 150])

In [ ]:
#compute the sum of all rows

In [ ]:
#Write your answer here in order not to overwrite the output, you wont be able to see the output if you write your code in the cell below
d_mat.sum(axis=1)

array([ 55, 155, 255, 355, 455])

In [ ]:
array([9, 9, 9, 9, 9])

# Data File Processing Using NumPy

In [ ]:
# Copyright 2016 Enthought, Inc. All Rights Reserved
"""
Wind Statistics
----------------

Topics: Using array methods over different axes, fancy indexing.

1. The data file 'wind.data' is in the week 9\Tutorial folder and has the following format:

        61  1  1 15.04 14.96 13.17  9.29 13.96  9.87 13.67 10.25 10.83 12.58 18.50 15.04
        61  1  2 14.71 16.88 10.83  6.50 12.62  7.67 11.50 10.04  9.79  9.67 17.54 13.83
        61  1  3 18.50 16.88 12.33 10.13 11.17  6.17 11.25  8.04  8.50  7.67 12.75 12.71

   The first three columns are year, month and day.  The
   remaining 12 columns are average windspeeds in knots at 12
   locations in Ireland on that day.

   Use the 'loadtxt' function from numpy to read the data into
   an array.

2. Calculate the min, max and mean windspeeds and standard deviation of the
   windspeeds over all the locations and all the times (a single set of numbers
   for the entire dataset).

3. Calculate the min, max and mean windspeeds and standard deviations of the
   windspeeds at each location over all the days (a different set of numbers
   for each location)

4. Calculate the min, max and mean windspeed and standard deviations of the
   windspeeds across all the locations at each day (a different set of numbers
   for each day)

5. Find the location which has the greatest windspeed on each day (an integer
   column number for each day).

6. Find the year, month and day on which the greatest windspeed was recorded.

7. Find the average windspeed in January for each location.

You should be able to perform all of these operations without using a for
loop or other looping construct.


In [ ]:
#tinyurl.com/3kbhprak

In [63]:
import numpy as np
from numpy import loadtxt
#f = open("wind.data", "r")
#a = np.array(f.read().split())
wind = loadtxt('wind.data')
data = wind[:, 3:]

print("Total:")
print("Minimum speed: ", data.min())
print("Maximum speed: ", data.max())
print("Mean speed: ", round(data.mean(),2))
print("Standard deviation: ", round(data.std(),2))

print("\nEach location: ")
print("Minimum speeds: ", data.min(axis=0))
print("Maximum speeds: ", data.max(axis=0))
print("Mean speeds: ", data.mean(axis=0))
print("Standard deviation: ", data.std(axis=0))

print("\nEach day: ")
print("Minimum speeds: ", data.min(axis=1))
print("Maximum speeds: ", data.max(axis=1))
print("Mean speeds: ", data.mean(axis=1))
print("Standard deviation: ", data.std(axis=1))

print("\nMaximum location of each day: ")
print(np.argmax(data, axis=1))

print("\nDate of highest record:")
loc2 = np.where(data == data.max())
year = wind[loc2[0], 0]
month = wind[loc2[0],1]
day = wind[loc2[0],2]
print(f"{int(year)+1900}/{int(month)}/{int(day)}")
print("To check: ", wind[loc2[0]])

#Q7
print("\nJanuary Statistics:")
jan_indices = wind[:,1] == 1
jan_data = data[jan_indices]
print('Mean: ', jan_data.mean(axis=0))


Total:
Minimum speed:  0.0
Maximum speed:  42.54
Mean speed:  10.23
Standard deviation:  5.6

Each location: 
Minimum speeds:  [0.67 0.21 1.5  0.   0.13 0.   0.   0.   0.   0.04 0.13 0.67]
Maximum speeds:  [35.8  33.37 33.84 28.46 37.54 26.16 30.37 31.08 25.88 28.21 42.38 42.54]
Mean speeds:  [12.36371463 10.64644813 11.66010344  6.30627472 10.45688013  7.09225434
  9.7968345   8.49442044  8.49581838  8.70726803 13.121007   15.59946152]
Standard deviation:  [5.61918301 5.26820081 5.00738377 3.60513309 4.93536333 3.96838126
 4.97689374 4.49865783 4.16746101 4.50327222 5.83459319 6.69734719]

Each day: 
Minimum speeds:  [9.29 6.5  6.17 ... 8.71 9.13 9.59]
Maximum speeds:  [18.5  17.54 18.5  ... 29.58 28.79 27.29]
Mean speeds:  [13.09666667 11.79833333 11.34166667 ... 14.89       15.3675
 15.4025    ]
Standard deviation:  [2.5773188  3.28972854 3.50543348 ... 5.51175108 5.30456427 5.45971172]

Maximum location of each day: 
[10 10  0 ... 11 11  2]

Date of highest record:
1966/12/2
To che